### ISSUE TITLE
Inconsistent type checking for sets of literal with implicit typing

### ISSUE THREAD

**Issue by JonathanPlasse** [microsoft/pyright#8647]
**Describe the bug**
Pyright is incorrectly flagging an error when passing variables with implicitly typed sets of literal values to a function expecting a set of literal types. This behavior is inconsistent with how Pyright handles explicitly typed sets and direct literal set arguments.

```
Argument of type "set[str]" cannot be assigned to parameter "_s" of type "set[Literal['a', 'b', 'c']]" in function "f"
  "set[str]" is incompatible with "set[Literal['a', 'b', 'c']]"
    Type parameter "_T@set" is invariant, but "str" is not the same as "Literal['a', 'b', 'c']"
    Consider switching from "set" to "Container" which is covariant
```

In my case, the suggestion to use Container instead of set is not always applicable.

**Code or Screenshots**

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2AULQCYnBTAAUA%2BgM4BcUHJMAbSKkKlAQCJy4gDRRxAIxlyAxuIC6agJRQAtAD4oAOTAoSXalEtQE5Dh1rVllWxygBBc1agMmrPpWBtfV5%2BIWIyKgkpWQUlcVUNTy8rEH4AVxAUEJgWTQcnFygAIRY3TSTLH2YWf0CK5PIePkFhCLFJOMUYhLUoAF4oAG8OgF8LZMtUmAys8nznOygAYVLy8crGatq1icn0zKHRh1ZhqRG8k46YxXPqDgBGJtDW0SjOuJ7%2Bw7O7gCYnlrhV5XORdFTqXoDU7vMZ%2Be4XGq-PIcADMX2hYw4ABZ0SDYrCaiiEdjNEA)

```python
from typing import Literal


def f(_s: set[Literal["a", "b", "c"]]) -> None:
    pass


class A:
    def f(self) -> set[Literal["a", "b", "c"]]:
        return set()


class B(A):
    def f(self):
        a: set[Literal["a", "b", "c"]] = {"a"}
        return a


class C(A):
    def f(self):
        return {"a"}  # error


f({"a"})
f({"a", "b"})
s1: set[Literal["a", "b", "c"]] = {"a"}
s2: set[Literal["a", "b", "c"]] = {"a", "b"}
f(s1)
f(s2)
s3 = {"a"}
s4 = {"a", "b"}
f(s3)  # error
f(s4)  # error
```

**VS Code extension or command-line**
Pyright 1.1.374

**Comment by erictraut** [comment]
Pyright is working correctly here. This isn't a bug.

Pyright (and other static type checkers) use a technique called [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types) when there is additional context in the form of an "expected type". This influences the inferences of expressions. In statements like `s3 = {"a"}` there is no such context, so pyright uses default inference rules. You can read about these default inference rules in the pyright documentation. In the case of `s3 = {"a"}`, the inferred type is `set[str]` rather than `set[Literal["a"]]` or `set[LiteralString]` or `set[object]`. If you would like to influence the inferred type, you can provide an explicit type annotation for `s3`, as in `s3: set[Literal["a"]] = {"a"}`.

Closing the issue because pyright is working as intended for the code sample.

**Comment by Andrej730** [comment]
@erictraut 

Is there a bidirectional type inference for return values? Met a similar issue but I thought in this case it should have consider the declared return type and this problem shouldn't occur.


```python
import typing


class Operator:
    def execute(
        self, context
    ) -> set[typing.Literal["RUNNING_MODAL", "CANCELLED", "FINISHED", "PASS_THROUGH", "INTERFACE"]]:
        ...


class Test(Operator):
    def execute(self, context):
        # Method "execute" overrides class "Operator" in an incompatible manner
        # Return type mismatch: base method returns type "set[Literal['RUNNING_MODAL', 'CANCELLED', 'FINISHED', 'PASS_THROUGH', 'INTERFACE']]", override returns type "set[str]"
        #     "set[str]" is incompatible with "set[Literal['RUNNING_MODAL', 'CANCELLED', 'FINISHED', 'PASS_THROUGH', 'INTERFACE']]"
        #     Type parameter "_T@set" is invariant, but "str" is not the same as "Literal['RUNNING_MODAL', 'CANCELLED', 'FINISHED', 'PASS_THROUGH', 'INTERFACE']"
        return {"FINISHED"}
```

**Comment by erictraut** [comment]
Yes, there is bidirectional type inference for return annotations. In this case, however, you have omitted the type annotation. Type inference doesn't use return annotations from other methods in other classes.
